crates/router/src/core/payment_methods/cards.rs

```diff
@@ -1394,13 +1727,110 @@ pub async fn add_payment_method(
 
 #[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
 #[instrument(skip_all)]
-pub async fn add_payment_method(
-    _state: &routes::SessionState,
-    _req: api::PaymentMethodCreate,
-    _merchant_account: &domain::MerchantAccount,
-    _key_store: &domain::MerchantKeyStore,
-) -> errors::RouterResponse<api::PaymentMethodResponse> {
-    todo!()
+pub async fn vault_payment_method(
+    state: &routes::SessionState,
+    pmd: &api::PaymentMethodCreateData,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+    customer_id: &id_type::CustomerId,
+) -> errors::RouterResult<payment_methods::VaultAddResp> {
+    let db = &*state.store;
+    let merchant_id = merchant_account.get_id();
+
+    // get fingerprint_id from locker
+    let fingerprint_id_from_locker =
+        get_fingerprint_id_from_locker(state, merchant_account, customer_id, pmd).await?;
+
+    // throw back error if payment method is duplicated
+    if db
+        .find_payment_method_by_fingerprint_id(
+            &(state.into()),
+            key_store,
+            &fingerprint_id_from_locker,
+        )
+        .await
+        .ok()
+        .is_some()
+    {
+        return Err(report!(errors::ApiErrorResponse::DuplicatePaymentMethod)
+            .attach_printable("Cannot vault duplicate payment method"));
+    }
+
+    let resp_from_locker =
+        vault_payment_method_in_locker(state, merchant_account, customer_id, pmd).await?;
+
+    Ok(resp_from_locker)
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub async fn get_fingerprint_id_from_locker(
+    state: &routes::SessionState,
+    merchant_account: &domain::MerchantAccount,
+    customer_id: &id_type::CustomerId,
+    pmd: &api::PaymentMethodCreateData,
+) -> errors::RouterResult<String> {
+    let card = match pmd {
+        api::PaymentMethodCreateData::Card(card) => Some(card.clone()),
+    }
+    .ok_or(errors::ApiErrorResponse::NotSupported {
+        message: "Payment method not supported".to_string(),
+    })?;
+
+    let key = card.card_number.clone().to_string();
```

Comment:
This can't be a specific function of card here, you need to implement a trait which should get key for any payment type